lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -466,210 +480,269 @@ class PullbackInfo {
+      linearMapStruct->setGenericParams(genericParams);
+      linearMapStruct->setGenericEnvironment(
+          genericSig->createGenericEnvironment());
     }
-    pullbackStruct->setBraces(loc);
+    linearMapStruct->setBraces(loc);
     computeAccessLevel(
-        pullbackStruct, original->getEffectiveSymbolLinkage());
-    pullbackStruct->computeType();
-    assert(pullbackStruct->hasInterfaceType());
-    file.addVisibleDecl(pullbackStruct);
+        linearMapStruct, original->getEffectiveSymbolLinkage());
+    linearMapStruct->computeType();
+    assert(linearMapStruct->hasInterfaceType());
+    file.addVisibleDecl(linearMapStruct);
     LLVM_DEBUG({
       auto &s = getADDebugStream();
-      s << "Pullback struct created for function @" << original->getName()
-        << " bb" << originalBB->getDebugID() << '\n';
-      pullbackStruct->print(s);
+      std::string structName;
+      switch (kind) {
+      case AutoDiffAssociatedFunctionKind::JVP:
+        structName = "Differential";
+        break;
+      case AutoDiffAssociatedFunctionKind::VJP:
+        structName = "Pullback";
+        break;
+      }
+      s << structName << " struct declaration created for function @"
+        << original->getName() << " bb" << originalBB->getDebugID() << '\n';
+      linearMapStruct->print(s);
       s << '\n';
     });
-    return pullbackStruct;
+    return linearMapStruct;
+  }
+
+  /// Add a linear map to the linear map struct.
+  VarDecl *addLinearMapDecl(SILInstruction *inst, SILType linearMapType) {
+    // IRGen requires decls to have AST types (not `SILFunctionType`), so we
+    // convert the `SILFunctionType` of the linear map to a `FunctionType` with
+    // the same parameters and results.
+    auto silFnTy = linearMapType.castTo<SILFunctionType>();
+    SmallVector<AnyFunctionType::Param, 8> params;
+    for (auto &param : silFnTy->getParameters())
+      params.push_back(AnyFunctionType::Param(param.getType()));
+    AnyFunctionType *astFnTy;
+    if (auto genSig = silFnTy->getGenericSignature())
+      astFnTy = GenericFunctionType::get(
+          genSig, params, silFnTy->getAllResultsType().getASTType());
+    else
+      astFnTy = FunctionType::get(
+          params, silFnTy->getAllResultsType().getASTType());
+
+    auto *origBB = inst->getParent();
+    auto *linMapStruct = getLinearMapStruct(origBB);
+    std::string linearMapName;
+    switch (kind) {
+    case swift::AutoDiffAssociatedFunctionKind::JVP:
+      linearMapName = "differential_" + llvm::itostr(linearMapValueMap.size());
+      break;
+    case swift::AutoDiffAssociatedFunctionKind::VJP:
+      linearMapName = "pullback_" + llvm::itostr(linearMapValueMap.size());
+      break;
+    }
+    auto *linearMapDecl = addVarDecl(linMapStruct, linearMapName, astFnTy);
+    linearMapValueMap.insert({inst, linearMapDecl});
+    return linearMapDecl;
   }
 
+  void prepareLinearMapStructDeclarations(
```

Comment:
Please add a doc comment clarifying that this method populates the fields of linear map structs. Or perhaps rename the method to something like `populateLinearMapStructDeclarationFields`.